// @munesoft/retryx/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// src/types.ts
var RetryError = class extends Error {
  constructor(message, errors) {
    super(message);
    this.name = "RetryError";
    this.errors = errors;
    this.attempts = errors.length;
    this.cause = errors[errors.length - 1];
  }
};
var AbortError = class extends Error {
  constructor(message = "Operation aborted") {
    super(message);
    this.name = "AbortError";
  }
};
var TimeoutError = class extends Error {
  constructor(message = "Operation timed out") {
    super(message);
    this.name = "TimeoutError";
  }
};

// src/backoff/strategies.ts
function computeBackoff(params) {
  const { attempt, delay, backoff, maxDelay, jitter, retryAfterMs } = params;
  if (typeof retryAfterMs === "number" && retryAfterMs >= 0) {
    return Math.min(retryAfterMs, maxDelay);
  }
  let raw;
  switch (backoff) {
    case "exponential":
      raw = delay * Math.pow(2, Math.min(attempt, 30));
      break;
    case "linear":
      raw = delay * (attempt + 1);
      break;
    case "fixed":
      raw = delay;
      break;
    default:
      raw = delay;
  }
  const capped = Math.min(raw, maxDelay);
  if (jitter) {
    return Math.floor(Math.random() * capped);
  }
  return capped;
}

// src/ai/detect.ts
var RETRYABLE_STATUS = /* @__PURE__ */ new Set([
  408,
  // Request Timeout
  409,
  // Conflict (often a retryable race in DB/AI APIs)
  425,
  // Too Early
  429,
  // Too Many Requests — the big one for rate limits
  500,
  // Internal Server Error
  502,
  // Bad Gateway
  503,
  // Service Unavailable
  504,
  // Gateway Timeout
  529
  // Site Overloaded (Anthropic uses this)
]);
var RETRYABLE_NETWORK_CODES = /* @__PURE__ */ new Set([
  "ECONNRESET",
  "ECONNREFUSED",
  "ETIMEDOUT",
  "EPIPE",
  "EAI_AGAIN",
  "ENOTFOUND",
  "ENETUNREACH",
  "EHOSTUNREACH",
  "EHOSTDOWN",
  "ECONNABORTED",
  // axios timeout
  "ERR_NETWORK",
  "UND_ERR_SOCKET",
  // undici
  "UND_ERR_CONNECT_TIMEOUT",
  "UND_ERR_HEADERS_TIMEOUT",
  "UND_ERR_BODY_TIMEOUT"
]);
function getStatus(error) {
  if (!error || typeof error !== "object") return void 0;
  const e = error;
  if (typeof e.status === "number") return e.status;
  if (typeof e.statusCode === "number") return e.statusCode;
  const response = e.response;
  if (response && typeof response === "object") {
    const r = response;
    if (typeof r.status === "number") return r.status;
    if (typeof r.statusCode === "number") return r.statusCode;
  }
  return void 0;
}
function getErrorCode(error) {
  if (!error || typeof error !== "object") return void 0;
  const e = error;
  if (typeof e.code === "string") return e.code;
  const cause = e.cause;
  if (cause && typeof cause === "object") {
    const c = cause;
    if (typeof c.code === "string") return c.code;
  }
  return void 0;
}
function isAbortError(error) {
  if (!error || typeof error !== "object") return false;
  const e = error;
  return e.name === "AbortError" || e.code === "ABORT_ERR" || e.code === "ERR_ABORTED";
}
function isRetryableError(error) {
  if (isAbortError(error)) return false;
  const status = getStatus(error);
  if (typeof status === "number") {
    if (RETRYABLE_STATUS.has(status)) return true;
    if (status >= 500 && status < 600) return true;
    if (status >= 400 && status < 500) return false;
  }
  const code = getErrorCode(error);
  if (code && RETRYABLE_NETWORK_CODES.has(code)) return true;
  if (error instanceof Error) {
    const msg = error.message.toLowerCase();
    if (msg.includes("fetch failed") || msg.includes("network error") || msg.includes("socket hang up") || msg.includes("econnreset") || msg.includes("etimedout")) {
      return true;
    }
  }
  return false;
}

// src/ai/retry-after.ts
function parseRetryAfter(value, now = Date.now()) {
  if (value === null || value === void 0) return void 0;
  if (typeof value === "number") {
    return value >= 0 ? value * 1e3 : void 0;
  }
  const trimmed = value.trim();
  if (!trimmed) return void 0;
  if (/^\d+(\.\d+)?$/.test(trimmed)) {
    const seconds = parseFloat(trimmed);
    if (!Number.isFinite(seconds) || seconds < 0) return void 0;
    return Math.round(seconds * 1e3);
  }
  const dateMs = Date.parse(trimmed);
  if (Number.isNaN(dateMs)) return void 0;
  const diff = dateMs - now;
  return diff > 0 ? diff : 0;
}
function readHeader(headers, name) {
  if (!headers) return void 0;
  if (headers instanceof Map) {
    const lower = name.toLowerCase();
    for (const [k, v] of headers.entries()) {
      if (typeof k === "string" && k.toLowerCase() === lower) {
        return v;
      }
    }
    return void 0;
  }
  if (typeof headers.get === "function") {
    try {
      const v = headers.get(name);
      return v ?? void 0;
    } catch {
    }
  }
  if (typeof headers === "object") {
    const obj = headers;
    const lower = name.toLowerCase();
    for (const k of Object.keys(obj)) {
      if (k.toLowerCase() === lower) {
        const v = obj[k];
        if (typeof v === "string" || typeof v === "number") return v;
        if (Array.isArray(v) && v.length > 0) {
          const first = v[0];
          if (typeof first === "string" || typeof first === "number")
            return first;
        }
      }
    }
  }
  return void 0;
}
function extractRetryAfterMs(error) {
  if (!error || typeof error !== "object") return void 0;
  const e = error;
  if (typeof e.retryAfter === "number") {
    return e.retryAfter >= 0 ? e.retryAfter * 1e3 : void 0;
  }
  const candidates = [
    e.headers,
    e.response?.headers
  ];
  for (const headers of candidates) {
    const raw = readHeader(headers, "retry-after");
    if (raw !== void 0) {
      const ms = parseRetryAfter(raw);
      if (ms !== void 0) return ms;
    }
  }
  return void 0;
}

// src/utils/abort.ts
function combineSignals(signals) {
  const real = signals.filter((s) => !!s);
  const controller = new AbortController();
  for (const s of real) {
    if (s.aborted) {
      const reason = s.reason;
      controller.abort(reason);
      return { signal: controller.signal, cancel: () => {
      } };
    }
  }
  const listeners = [];
  for (const s of real) {
    const handler = () => {
      const reason = s.reason;
      controller.abort(reason);
    };
    s.addEventListener("abort", handler, { once: true });
    listeners.push({ signal: s, handler });
  }
  const cancel = () => {
    for (const { signal, handler } of listeners) {
      signal.removeEventListener("abort", handler);
    }
  };
  return { signal: controller.signal, cancel };
}

// src/utils/sleep.ts
function sleep(ms, signal) {
  if (ms <= 0) {
    if (signal?.aborted) return Promise.reject(new AbortError());
    return Promise.resolve();
  }
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new AbortError());
      return;
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(new AbortError());
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

// src/core/engine.ts
function resolveOptions(opts = {}) {
  const ai = opts.ai === true;
  return {
    retries: opts.retries ?? (ai ? 5 : 3),
    delay: opts.delay ?? (ai ? 500 : 200),
    backoff: ai ? "exponential" : opts.backoff ?? "exponential",
    jitter: opts.jitter ?? true,
    maxDelay: opts.maxDelay ?? 3e4,
    totalTimeout: opts.totalTimeout ?? 0,
    timeout: opts.timeout ?? 0,
    ai,
    retry: opts.retry,
    signal: opts.signal,
    fallback: opts.fallback,
    onRetry: opts.onRetry,
    onSuccess: opts.onSuccess,
    onFail: opts.onFail
  };
}
async function shouldRetry(error, attempt, resolved) {
  if (isAbortError(error)) return false;
  if (resolved.retry) {
    return await resolved.retry(error, attempt);
  }
  if (resolved.ai) {
    return isRetryableError(error);
  }
  return true;
}
async function retryx(fn, options) {
  const opts = resolveOptions(options);
  if (typeof fn !== "function") {
    throw new TypeError("retryx: first argument must be a function");
  }
  if (opts.retries < 0 || !Number.isFinite(opts.retries)) {
    throw new RangeError("retryx: `retries` must be a non-negative number");
  }
  const startTime = Date.now();
  const errors = [];
  const totalDeadlineController = new AbortController();
  let totalDeadlineTimer;
  if (opts.totalTimeout > 0) {
    totalDeadlineTimer = setTimeout(() => {
      totalDeadlineController.abort(new TimeoutError("Total timeout exceeded"));
    }, opts.totalTimeout);
    if (typeof totalDeadlineTimer === "object" && totalDeadlineTimer !== null) {
      const t = totalDeadlineTimer;
      t.unref?.();
    }
  }
  try {
    for (let attempt = 0; attempt <= opts.retries; attempt++) {
      const perAttemptController = new AbortController();
      let perAttemptTimer;
      if (opts.timeout > 0) {
        perAttemptTimer = setTimeout(() => {
          perAttemptController.abort(new TimeoutError("Attempt timed out"));
        }, opts.timeout);
        if (typeof perAttemptTimer === "object" && perAttemptTimer !== null) {
          const t = perAttemptTimer;
          t.unref?.();
        }
      }
      const { signal: attemptSignal, cancel: cancelAttemptSignal } = combineSignals([
        opts.signal,
        totalDeadlineController.signal,
        perAttemptController.signal
      ]);
      if (opts.signal?.aborted) {
        throw new AbortError("Aborted before attempt");
      }
      if (totalDeadlineController.signal.aborted) {
        throw new TimeoutError("Total timeout exceeded");
      }
      try {
        const result = await fn({ attempt, signal: attemptSignal });
        if (perAttemptTimer) clearTimeout(perAttemptTimer);
        cancelAttemptSignal();
        if (opts.onSuccess) {
          try {
            await opts.onSuccess(result, attempt);
          } catch {
          }
        }
        return result;
      } catch (error) {
        if (perAttemptTimer) clearTimeout(perAttemptTimer);
        cancelAttemptSignal();
        if (opts.signal?.aborted) {
          errors.push(error);
          throw new AbortError("Aborted by signal");
        }
        if (totalDeadlineController.signal.aborted) {
          errors.push(error);
          throw new TimeoutError("Total timeout exceeded");
        }
        let normalizedError = error;
        if (perAttemptController.signal.aborted && isAbortError(error)) {
          normalizedError = new TimeoutError(
            `Attempt ${attempt} timed out after ${opts.timeout}ms`
          );
        }
        errors.push(normalizedError);
        if (attempt >= opts.retries) {
          throw normalizedError;
        }
        const retryable = await shouldRetry(normalizedError, attempt, opts);
        if (!retryable) {
          throw normalizedError;
        }
        const retryAfterMs = opts.ai ? extractRetryAfterMs(normalizedError) : void 0;
        let nextDelay = computeBackoff({
          attempt,
          delay: opts.delay,
          backoff: opts.backoff ?? "exponential",
          maxDelay: opts.maxDelay,
          jitter: opts.jitter,
          retryAfterMs
        });
        if (opts.totalTimeout > 0) {
          const elapsed = Date.now() - startTime;
          const remaining = opts.totalTimeout - elapsed;
          if (remaining <= 0) {
            throw new TimeoutError("Total timeout exceeded");
          }
          if (nextDelay > remaining) nextDelay = remaining;
        }
        if (opts.onRetry) {
          let hookResult = void 0;
          try {
            hookResult = await opts.onRetry(
              normalizedError,
              attempt,
              nextDelay
            );
          } catch {
          }
          if (hookResult === false) {
            throw normalizedError;
          }
        }
        const { signal: sleepSignal, cancel: cancelSleepSignal } = combineSignals([opts.signal, totalDeadlineController.signal]);
        try {
          await sleep(nextDelay, sleepSignal);
        } catch (sleepErr) {
          cancelSleepSignal();
          if (opts.signal?.aborted) throw new AbortError("Aborted by signal");
          if (totalDeadlineController.signal.aborted) {
            throw new TimeoutError("Total timeout exceeded");
          }
          throw sleepErr;
        }
        cancelSleepSignal();
      }
    }
    throw new RetryError("retryx: exhausted retries", errors);
  } catch (finalError) {
    if (opts.onFail) {
      try {
        await opts.onFail(finalError, errors.length);
      } catch {
      }
    }
    if (opts.fallback !== void 0) {
      if (typeof opts.fallback === "function") {
        return await opts.fallback(
          finalError
        );
      }
      return opts.fallback;
    }
    if (finalError instanceof TimeoutError || finalError instanceof AbortError) {
      throw finalError;
    }
    if (errors.length > 1) {
      throw new RetryError(
        `retryx: failed after ${errors.length} attempts: ${finalError instanceof Error ? finalError.message : String(finalError)}`,
        errors
      );
    }
    throw finalError;
  } finally {
    if (totalDeadlineTimer) clearTimeout(totalDeadlineTimer);
  }
}

// src/index.ts
var index_default = retryx;

exports.AbortError = AbortError;
exports.RetryError = RetryError;
exports.TimeoutError = TimeoutError;
exports.computeBackoff = computeBackoff;
exports.default = index_default;
exports.extractRetryAfterMs = extractRetryAfterMs;
exports.getErrorCode = getErrorCode;
exports.getStatus = getStatus;
exports.isAbortError = isAbortError;
exports.isRetryableError = isRetryableError;
exports.parseRetryAfter = parseRetryAfter;
exports.retryx = retryx;
