// @munesoft/idx/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default,
  idx: () => idx,
  randomId: () => randomId,
  readableId: () => readableId,
  timeId: () => timeId
});
module.exports = __toCommonJS(index_exports);

// src/random.ts
var isBrowser = typeof globalThis !== "undefined" && typeof globalThis.crypto !== "undefined" && typeof globalThis.crypto.getRandomValues === "function";
var POOL_SIZE = 512;
var pool = new Uint8Array(POOL_SIZE);
var poolOffset = POOL_SIZE;
function fillPool() {
  if (isBrowser) {
    globalThis.crypto.getRandomValues(pool);
  } else {
    try {
      const buf = require("crypto").randomBytes(POOL_SIZE);
      pool.set(buf);
    } catch {
      for (let i = 0; i < POOL_SIZE; i++) {
        pool[i] = Math.floor(Math.random() * 256);
      }
      if (typeof console !== "undefined") {
        console.warn("[idx] Using Math.random fallback \u2014 not cryptographically secure");
      }
    }
  }
  poolOffset = 0;
}
function getRandomBytes(length) {
  const out = new Uint8Array(length);
  let filled = 0;
  while (filled < length) {
    if (poolOffset >= POOL_SIZE) fillPool();
    const take = Math.min(POOL_SIZE - poolOffset, length - filled);
    out.set(pool.subarray(poolOffset, poolOffset + take), filled);
    poolOffset += take;
    filled += take;
  }
  return out;
}

// src/encode.ts
var BASE62 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
var B62_LEN = 62;
var B62_LOOKUP = new Array(256);
for (let i = 0; i < 256; i++) B62_LOOKUP[i] = BASE62[i % B62_LEN];
function encodeBase62(num) {
  if (num === 0) return "0";
  let result = "";
  let n = num;
  while (n > 0) {
    result = BASE62[n % B62_LEN] + result;
    n = Math.floor(n / B62_LEN);
  }
  return result;
}
function encodeBase62Padded(num, length) {
  const raw = encodeBase62(num);
  return raw.padStart(length, "0");
}
function bytesToBase62(bytes) {
  const len = bytes.length;
  const chars = new Array(len);
  for (let i = 0; i < len; i++) chars[i] = B62_LOOKUP[bytes[i]];
  return chars.join("");
}

// src/time.ts
var TS_LENGTH = 8;
function timeId(options = {}) {
  const randomLength = options.randomLength ?? 8;
  const ts = Date.now();
  const tsPart = encodeBase62Padded(ts, TS_LENGTH);
  const randPart = bytesToBase62(getRandomBytes(randomLength));
  return tsPart + randPart;
}
function extractTimestamp(id) {
  if (id.length < TS_LENGTH) return null;
  const tsPart = id.slice(0, TS_LENGTH);
  let n = 0;
  for (let i = 0; i < tsPart.length; i++) {
    const c = tsPart.charCodeAt(i);
    let digit;
    if (c >= 48 && c <= 57) digit = c - 48;
    else if (c >= 65 && c <= 90) digit = c - 65 + 10;
    else if (c >= 97 && c <= 122) digit = c - 97 + 36;
    else return null;
    n = n * 62 + digit;
  }
  return n;
}

// src/readable.ts
var ADJECTIVES = [
  "blue",
  "red",
  "gold",
  "dark",
  "fast",
  "slow",
  "cold",
  "warm",
  "bold",
  "calm",
  "deep",
  "flat",
  "free",
  "grey",
  "hard",
  "holy",
  "huge",
  "keen",
  "kind",
  "late",
  "lean",
  "lone",
  "loud",
  "mild",
  "neat",
  "pale",
  "pink",
  "pure",
  "rich",
  "safe",
  "slim",
  "soft",
  "tall",
  "tame",
  "thin",
  "tidy",
  "tiny",
  "true",
  "vast",
  "wide",
  "wise",
  "wild",
  "worn",
  "acid",
  "aged",
  "airy",
  "bare",
  "best",
  "clean",
  "crisp"
];
var NOUNS = [
  "fox",
  "cat",
  "dog",
  "owl",
  "elk",
  "bee",
  "ant",
  "bat",
  "cod",
  "emu",
  "jay",
  "ram",
  "yak",
  "cub",
  "cub",
  "lake",
  "hill",
  "rock",
  "wave",
  "sand",
  "leaf",
  "tree",
  "vine",
  "moss",
  "fog",
  "snow",
  "rain",
  "wind",
  "fire",
  "ash",
  "iron",
  "gold",
  "coal",
  "clay",
  "gem",
  "reef",
  "cave",
  "peak",
  "vale",
  "glen",
  "hawk",
  "wolf",
  "bear",
  "crow",
  "dove",
  "swan",
  "frog",
  "newt",
  "mole",
  "wren"
];
function readableId(options = {}) {
  const adjs = options.adjectives ?? ADJECTIVES;
  const nouns = options.nouns ?? NOUNS;
  const max = options.maxNumber ?? 99;
  const sep = options.separator ?? "-";
  const bytes = getRandomBytes(3);
  const adj = adjs[bytes[0] % adjs.length];
  const noun = nouns[bytes[1] % nouns.length];
  const num = bytes[2] % (max + 1);
  return `${adj}${sep}${noun}${sep}${num}`;
}

// src/utils.ts
var BASE62_CHARSET = new Set(BASE62.split(""));
function isBase62(str) {
  for (let i = 0; i < str.length; i++) {
    if (!BASE62_CHARSET.has(str[i])) return false;
  }
  return true;
}
function isReadableId(id) {
  return /^[a-z]+-[a-z]+-\d+$/.test(id);
}
function isTimeId(id) {
  return id.length >= 8 && isBase62(id) && !isReadableId(id);
}

// src/index.ts
var DEFAULT_LENGTH = 10;
function randomId(length = DEFAULT_LENGTH) {
  if (length <= 0) return "";
  return bytesToBase62(getRandomBytes(length));
}
function idx(arg) {
  if (arg === void 0) return randomId();
  if (typeof arg === "number") return randomId(arg);
  return randomId(arg.length);
}
idx.time = function(options) {
  return timeId(options);
};
idx.readable = function(options) {
  return readableId(options);
};
idx.batch = function(count, arg) {
  if (count <= 0) return [];
  const out = new Array(count);
  for (let i = 0; i < count; i++) {
    out[i] = idx(arg);
  }
  return out;
};
idx.meta = function(id) {
  if (isReadableId(id)) {
    const parts = id.split("-");
    return {
      type: "readable",
      segments: parts,
      raw: id,
      length: id.length
    };
  }
  if (isTimeId(id) && id.length >= 16) {
    const ts = extractTimestamp(id);
    return {
      type: "time",
      timestamp: ts ?? void 0,
      segments: [id.slice(0, 8), id.slice(8)],
      raw: id,
      length: id.length
    };
  }
  return {
    type: "random",
    raw: id,
    length: id.length
  };
};
var index_default = idx;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  idx,
  randomId,
  readableId,
  timeId
});
