// @munesoft/api-normalizer/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toString = exports.toNumber = exports.toDate = exports.toBoolean = exports.isPlainObject = exports.isEmpty = exports.get = exports.deepClone = exports.autoCoerce = void 0;
// @munesoft/api-normalizer — utils barrel
var helpers_1 = require("./helpers");
Object.defineProperty(exports, "autoCoerce", { enumerable: true, get: function () { return helpers_1.autoCoerce; } });
Object.defineProperty(exports, "deepClone", { enumerable: true, get: function () { return helpers_1.deepClone; } });
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return helpers_1.get; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return helpers_1.isEmpty; } });
Object.defineProperty(exports, "isPlainObject", { enumerable: true, get: function () { return helpers_1.isPlainObject; } });
Object.defineProperty(exports, "toBoolean", { enumerable: true, get: function () { return helpers_1.toBoolean; } });
Object.defineProperty(exports, "toDate", { enumerable: true, get: function () { return helpers_1.toDate; } });
Object.defineProperty(exports, "toNumber", { enumerable: true, get: function () { return helpers_1.toNumber; } });
Object.defineProperty(exports, "toString", { enumerable: true, get: function () { return helpers_1.toString; } });
