// @munesoft/envx/src/loader.js
import { loadFile } from "./parser.js";
import { validate } from "./validator.js";
import { loadConfig } from "./configLoader.js";

/**
 * Formats and prints a validation failure message, then exits.
 *
 * @param {string[]} errors
 */
function failWithErrors(errors) {
  console.error("\n❌ ENV VALIDATION FAILED\n");
  for (const err of errors) {
    console.error(`  - ${err}`);
  }
  console.error("");
  process.exit(1);
}

/**
 * Main envx loader.
 *
 * 1. Reads .env file (or custom path)
 * 2. Merges with process.env (process.env takes precedence unless overrideEnv=true)
 * 3. Loads envx.config.js schema (if present)
 * 4. Validates and coerces all variables
 * 5. Applies validated values back to process.env
 * 6. Optionally prints debug output
 *
 * @param {LoadOptions} [options]
 * @returns {Promise<Record<string, any>>}
 */
export async function load(options = {}) {
  const {
    path,
    override = false,
    debug = process.env.ENVX_DEBUG === "true",
    schema: inlineSchema,
    strict: inlineStrict,
    silent = false,
  } = options;

  // Step 1: Parse .env file
  const fileVars = loadFile(path);

  // Step 2: Merge with process.env
  // By default, existing process.env values win (they may come from the real environment)
  const merged = { ...fileVars };
  for (const [key, val] of Object.entries(process.env)) {
    if (override || merged[key] === undefined) {
      merged[key] = val;
    } else if (!override) {
      merged[key] = process.env[key] ?? merged[key];
    }
  }

  // Step 3: Load schema from config file (unless inline schema provided)
  let schema = inlineSchema ?? null;
  let strict = inlineStrict ?? false;

  if (!schema) {
    try {
      const config = await loadConfig();
      if (config) {
        schema = config.schema;
        strict = config.strict ?? false;
      }
    } catch (err) {
      if (!silent) {
        console.warn(`⚠️  envx: ${err.message}`);
      }
    }
  }

  // Step 4: Validate and coerce
  let finalVars = { ...merged };

  if (schema && Object.keys(schema).length > 0) {
    const { result, errors } = validate(merged, schema, strict);

    if (errors.length > 0) {
      failWithErrors(errors);
    }

    // Merge coerced values back — coerced typed values stored separately
    finalVars = { ...merged, ...result };
  }

  // Step 5: Apply string-coerced vars back to process.env
  for (const [key, val] of Object.entries(fileVars)) {
    if (override || process.env[key] === undefined) {
      process.env[key] = String(val);
    }
  }

  // Step 6: Debug output
  if (debug) {
    console.log("\nenvx loaded:");
    for (const [key, val] of Object.entries(fileVars)) {
      const displayVal =
        key.toLowerCase().includes("secret") ||
        key.toLowerCase().includes("password") ||
        key.toLowerCase().includes("key") ||
        key.toLowerCase().includes("token")
          ? "***"
          : val;
      console.log(`  ${key}=${displayVal}`);
    }
    console.log("");
  }

  return finalVars;
}

/**
 * Synchronous-style loader (for use in envx/config side-effect import).
 * Runs load() and captures the result for later access.
 */
let _loadPromise = null;
let _resolvedVars = null;

export function loadSync(options = {}) {
  _loadPromise = load(options).then((vars) => {
    _resolvedVars = vars;
  });
  return _loadPromise;
}

export function getLoadPromise() {
  return _loadPromise;
}

export function getResolvedVars() {
  return _resolvedVars;
}
