// @munesoft/loopx/dist/index.js
// src/step/index.ts
function createStep(iteration, state, signal, data, internals) {
  const step = {
    iteration,
    state,
    signal,
    data,
    get stopped() {
      return internals.isStopped();
    },
    stop(reason) {
      internals.requestStop(reason);
    },
    next(d) {
      internals.setNextData(d);
    }
  };
  return step;
}

// src/hooks/index.ts
async function fireOnStart(hooks) {
  if (hooks.onStart) await hooks.onStart();
}
async function fireOnStep(hooks, step) {
  if (hooks.onStep) await hooks.onStep(step);
}
async function fireOnStop(hooks, reason, detail) {
  if (hooks.onStop) await hooks.onStop(reason, detail);
}
async function fireOnError(hooks, error, step) {
  if (!hooks.onError) return false;
  const result = await hooks.onError(error, step);
  return Boolean(result);
}
async function fireOnComplete(hooks, result) {
  if (hooks.onComplete) await hooks.onComplete(result);
}

// src/utils/index.ts
function delay(ms, signal) {
  if (ms <= 0) return Promise.resolve();
  if (signal?.aborted) return Promise.resolve();
  return new Promise((resolve) => {
    const timer = setTimeout(() => {
      if (signal) signal.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      resolve();
    };
    if (signal) signal.addEventListener("abort", onAbort, { once: true });
  });
}
function stableHash(value) {
  const seen = /* @__PURE__ */ new WeakSet();
  const walk = (v) => {
    if (v === null) return "null";
    if (v === void 0) return "undefined";
    const t = typeof v;
    if (t === "string") return `s:${v}`;
    if (t === "number") return `n:${String(v)}`;
    if (t === "boolean") return `b:${String(v)}`;
    if (t === "bigint") return `i:${String(v)}`;
    if (t === "function" || t === "symbol") return `${t}`;
    if (v instanceof Date) return `d:${v.getTime()}`;
    if (v instanceof RegExp) return `r:${v.toString()}`;
    if (typeof v === "object") {
      const obj = v;
      if (seen.has(obj)) return "[cycle]";
      seen.add(obj);
      if (Array.isArray(v)) {
        return `[${v.map(walk).join(",")}]`;
      }
      const keys = Object.keys(v).sort();
      const entries = keys.map((k) => `${k}:${walk(v[k])}`);
      return `{${entries.join(",")}}`;
    }
    return String(v);
  };
  return walk(value);
}
var DEFAULT_MAX_ITERATIONS = 1e3;
var DEFAULT_STAGNATION_THRESHOLD = 3;

// src/core/runner.ts
var noopCoordination = {
  isPaused: () => false,
  waitWhilePaused: async () => {
  },
  externallyStopped: () => ({ stopped: false })
};
async function runLoop(fn, options = {}, coordination = noopCoordination) {
  const startTime = Date.now();
  const state = { ...options.initialState ?? {} };
  const maxIterations = options.maxIterations ?? DEFAULT_MAX_ITERATIONS;
  const stagnationThreshold = options.stagnationThreshold ?? DEFAULT_STAGNATION_THRESHOLD;
  const abortController = new AbortController();
  let externalAbortHandler = null;
  if (options.signal) {
    if (options.signal.aborted) {
      abortController.abort();
    } else {
      externalAbortHandler = () => {
        stopRequested = true;
        stopReason = stopReason ?? "aborted";
        abortController.abort();
      };
      options.signal.addEventListener("abort", externalAbortHandler, { once: true });
    }
  }
  let timeoutHandle = null;
  if (options.timeout && options.timeout > 0 && Number.isFinite(options.timeout)) {
    timeoutHandle = setTimeout(() => {
      if (!stopRequested) {
        stopRequested = true;
        stopReason = "timeout";
      }
      abortController.abort();
    }, options.timeout);
    const handle = timeoutHandle;
    if (typeof handle.unref === "function") {
      handle.unref();
    }
  }
  let stopRequested = false;
  let stopReason = null;
  let stopDetail;
  const requestStop = (reason, detail) => {
    if (stopRequested) return;
    stopRequested = true;
    stopReason = reason;
    stopDetail = detail;
    if (!abortController.signal.aborted) abortController.abort();
  };
  let lastHash = null;
  let repeatCount = 0;
  try {
    await fireOnStart(options);
  } catch (e) {
    cleanup();
    throw e;
  }
  let iteration = 0;
  let nextData = void 0;
  let loopError = void 0;
  try {
    while (true) {
      const ext = coordination.externallyStopped();
      if (ext.stopped && !stopRequested) {
        requestStop("controller_stop", ext.reason);
      }
      if (coordination.isPaused() && !stopRequested) {
        await coordination.waitWhilePaused(abortController.signal);
      }
      if (stopRequested) break;
      if (abortController.signal.aborted) break;
      if (iteration >= maxIterations) {
        requestStop("max_iterations", `Reached cap of ${maxIterations}`);
        break;
      }
      const currentData = nextData;
      nextData = void 0;
      let stepStopRequested = false;
      let stepStopReasonText;
      let stepNextData = void 0;
      const step = createStep(
        iteration,
        state,
        abortController.signal,
        currentData,
        {
          requestStop: (r) => {
            stepStopRequested = true;
            stepStopReasonText = r;
          },
          setNextData: (d) => {
            stepNextData = d;
          },
          isStopped: () => stepStopRequested || stopRequested
        }
      );
      let attempt = 0;
      const maxAttempts = (options.retry ?? 0) + 1;
      let succeeded = false;
      let lastError;
      while (attempt < maxAttempts && !succeeded) {
        try {
          if (attempt === 0) {
            await fireOnStep(options, step);
          }
          await fn(step);
          succeeded = true;
        } catch (err) {
          lastError = err;
          attempt += 1;
          if (attempt >= maxAttempts) {
            const suppressed = await fireOnError(options, err, step);
            if (suppressed) {
              succeeded = true;
              lastError = void 0;
            } else {
              loopError = err;
              requestStop("error", err instanceof Error ? err.message : String(err));
            }
          } else {
            if (options.delay && options.delay > 0) {
              await delay(options.delay, abortController.signal);
            }
            if (abortController.signal.aborted || stopRequested) break;
          }
        }
      }
      if (!succeeded && stopRequested) break;
      if (stepStopRequested) {
        requestStop("manual", stepStopReasonText);
      }
      if (stepNextData !== void 0) {
        nextData = stepNextData;
      }
      iteration += 1;
      if (loopError !== void 0) break;
      if (options.ai && stepNextData !== void 0) {
        const hash = stableHash(stepNextData);
        if (hash === lastHash) {
          repeatCount += 1;
          if (repeatCount >= stagnationThreshold) {
            requestStop(
              "stagnation",
              `Output unchanged for ${stagnationThreshold} consecutive iterations`
            );
            break;
          }
        } else {
          lastHash = hash;
          repeatCount = 1;
        }
      }
      if (options.stop) {
        try {
          const shouldStop = await options.stop(step);
          if (shouldStop) {
            requestStop("stop_condition");
            break;
          }
        } catch (err) {
          loopError = err;
          requestStop("error", err instanceof Error ? err.message : String(err));
          break;
        }
      }
      if (stopRequested || abortController.signal.aborted) break;
      if (options.delay && options.delay > 0) {
        await delay(options.delay, abortController.signal);
        if (stopRequested || abortController.signal.aborted) break;
      }
    }
  } catch (err) {
    loopError = err;
    if (!stopRequested) requestStop("error", err instanceof Error ? err.message : String(err));
  } finally {
    cleanup();
  }
  if (!stopReason) {
    stopReason = "completed";
  }
  const result = {
    iterations: iteration,
    reason: stopReason,
    detail: stopDetail,
    state,
    elapsedMs: Date.now() - startTime,
    error: loopError
  };
  try {
    await fireOnStop(options, stopReason, stopDetail);
  } catch {
  }
  try {
    await fireOnComplete(options, result);
  } catch {
  }
  if (loopError !== void 0) {
    throw loopError;
  }
  return result;
  function cleanup() {
    if (timeoutHandle) {
      clearTimeout(timeoutHandle);
      timeoutHandle = null;
    }
    if (externalAbortHandler && options.signal) {
      options.signal.removeEventListener("abort", externalAbortHandler);
      externalAbortHandler = null;
    }
  }
}

// src/control/index.ts
function createController() {
  let paused = false;
  let running = false;
  let stopFlag = { stopped: false };
  let pauseWaiters = [];
  const coordination = {
    isPaused: () => paused,
    waitWhilePaused: (signal) => {
      if (!paused) return Promise.resolve();
      return new Promise((resolve) => {
        const onResume = () => resolve();
        const onAbort = () => {
          pauseWaiters = pauseWaiters.filter((w) => w !== onResume);
          resolve();
        };
        pauseWaiters.push(onResume);
        if (signal.aborted) {
          onAbort();
        } else {
          signal.addEventListener("abort", onAbort, { once: true });
        }
      });
    },
    externallyStopped: () => stopFlag
  };
  const controller2 = {
    async run(fn, options) {
      if (running) {
        throw new Error("Controller is already running a loop. Wait for it to finish before calling run() again.");
      }
      running = true;
      paused = false;
      stopFlag = { stopped: false };
      pauseWaiters = [];
      try {
        return await runLoop(fn, options, coordination);
      } finally {
        running = false;
        paused = false;
        const drain = pauseWaiters;
        pauseWaiters = [];
        for (const w of drain) w();
      }
    },
    pause() {
      if (!running) return;
      paused = true;
    },
    resume() {
      if (!paused) return;
      paused = false;
      const drain = pauseWaiters;
      pauseWaiters = [];
      for (const w of drain) w();
    },
    stop(reason) {
      stopFlag = { stopped: true, reason };
      if (paused) {
        paused = false;
        const drain = pauseWaiters;
        pauseWaiters = [];
        for (const w of drain) w();
      }
    },
    get paused() {
      return paused;
    },
    get running() {
      return running;
    }
  };
  return controller2;
}

// src/index.ts
function loopx(fn, options) {
  return runLoop(fn, options);
}
function controller() {
  return createController();
}
loopx.controller = controller;
var src_default = loopx;

export { controller, src_default as default, loopx };
