// @munesoft/api-normalizer/dist/mapper.js
"use strict";
// ─────────────────────────────────────────────────────────────────────────────
// @munesoft/api-normalizer — Mapper Engine
// ─────────────────────────────────────────────────────────────────────────────
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapObject = mapObject;
exports.inferSchema = inferSchema;
const helpers_1 = require("./utils/helpers");
// ─── Internal Helpers ────────────────────────────────────────────────────────
/** Normalise a SchemaField to the canonical FieldMapping shape */
function toFieldMapping(field) {
    if (Array.isArray(field)) {
        return { keys: field };
    }
    return field;
}
/**
 * Resolve the first matching key from `candidates` found in `data`.
 * Supports dot-notation paths for nested access.
 * Returns `{ key, value }` or `{ key: null, value: undefined }`.
 */
function resolveKey(data, candidates) {
    for (const candidate of candidates) {
        const value = (0, helpers_1.get)(data, candidate);
        if (!(0, helpers_1.isEmpty)(value)) {
            return { key: candidate, value };
        }
    }
    return { key: null, value: undefined };
}
/** Apply type coercion based on the CoercionType hint */
function applyCoercion(value, type) {
    switch (type) {
        case "string":
            return (0, helpers_1.toString)(value);
        case "number": {
            const n = (0, helpers_1.toNumber)(value);
            return isNaN(n) ? value : n;
        }
        case "boolean":
            return (0, helpers_1.toBoolean)(value);
        case "date":
            return (0, helpers_1.toDate)(value);
        case "auto":
            return (0, helpers_1.autoCoerce)(value);
        default:
            return value;
    }
}
/**
 * Map a single plain object against a schema.
 * Returns the normalized object plus diff and validation metadata.
 */
function mapObject(data, schema, options = {}) {
    const normalized = {};
    const diff = [];
    const missingRequired = [];
    for (const [outputKey, rawField] of Object.entries(schema)) {
        const mapping = toFieldMapping(rawField);
        const { keys, default: defaultValue, type, required, transform } = mapping;
        const { key: sourceKey, value: rawValue } = resolveKey(data, keys);
        let usedDefault = false;
        let resolvedValue = rawValue;
        // Fall back to default when nothing resolved
        if ((0, helpers_1.isEmpty)(resolvedValue)) {
            if (!(0, helpers_1.isEmpty)(defaultValue)) {
                resolvedValue = defaultValue;
                usedDefault = true;
            }
            else if (required) {
                missingRequired.push(outputKey);
                resolvedValue = null;
            }
            else {
                resolvedValue = null;
            }
        }
        // Type coercion
        let coerced = false;
        if (options.coerceTypes && type && !(0, helpers_1.isEmpty)(resolvedValue) && !usedDefault) {
            const coercedValue = applyCoercion(resolvedValue, type);
            if (coercedValue !== resolvedValue) {
                coerced = true;
                resolvedValue = coercedValue;
            }
        }
        else if (type && !(0, helpers_1.isEmpty)(resolvedValue) && !usedDefault) {
            // Always honour explicit type hint even without global coerceTypes flag
            const coercedValue = applyCoercion(resolvedValue, type);
            if (coercedValue !== resolvedValue) {
                coerced = true;
                resolvedValue = coercedValue;
            }
        }
        // Custom transform
        if (typeof transform === "function" && !(0, helpers_1.isEmpty)(resolvedValue)) {
            resolvedValue = transform(resolvedValue, data);
        }
        normalized[outputKey] = resolvedValue;
        if (options.logDiff) {
            diff.push({
                outputKey,
                sourceKey,
                rawValue,
                normalizedValue: resolvedValue,
                coerced,
                usedDefault,
            });
        }
    }
    return { normalized, diff, missingRequired };
}
// ─── Schema Inference ─────────────────────────────────────────────────────────
/**
 * Infer a basic schema from a sample object.
 * Each top-level key becomes a schema entry pointing to itself.
 * Nested objects use dot-notation paths.
 */
function inferSchema(data, prefix = "", depth = 0) {
    const schema = {};
    if (!(0, helpers_1.isPlainObject)(data) || depth > 4)
        return schema;
    for (const [key, value] of Object.entries(data)) {
        const fullKey = prefix ? `${prefix}.${key}` : key;
        const outputKey = key; // top-level key as output name
        if ((0, helpers_1.isPlainObject)(value) && depth < 4) {
            // Recurse one level for nested objects
            const nested = inferSchema(value, fullKey, depth + 1);
            Object.assign(schema, nested);
        }
        else {
            schema[outputKey] = { keys: [fullKey] };
        }
    }
    return schema;
}
