// @munesoft/api-normalizer/dist/normalize.js
"use strict";
// ─────────────────────────────────────────────────────────────────────────────
// @munesoft/api-normalizer — normalize()
// ─────────────────────────────────────────────────────────────────────────────
Object.defineProperty(exports, "__esModule", { value: true });
exports.inferSchema = void 0;
exports.normalize = normalize;
const mapper_1 = require("./mapper");
const helpers_1 = require("./utils/helpers");
// ─── Single Object ────────────────────────────────────────────────────────────
function normalizeSingle(data, schema, options) {
    const { normalized, diff, missingRequired } = (0, mapper_1.mapObject)(data, schema, options);
    // Strict mode: bail on missing required fields
    if (options.strict && missingRequired.length > 0) {
        const error = {
            success: false,
            error: `Missing required fields: ${missingRequired.join(", ")}`,
            missingFields: missingRequired,
        };
        return error;
    }
    // Apply plugins
    let result = normalized;
    if (options.plugins) {
        for (const plugin of options.plugins) {
            try {
                result = plugin.transform(result, data);
            }
            catch (err) {
                const message = err instanceof Error ? err.message : String(err);
                console.warn(`[api-normalizer] Plugin "${plugin.name}" threw: ${message}`);
            }
        }
    }
    return Object.assign(Object.assign({ success: true, data: result }, (options.inferSchema ? { _schema: (0, mapper_1.inferSchema)(data) } : {})), (options.logDiff ? { _diff: diff } : {}));
}
// ─── Array ────────────────────────────────────────────────────────────────────
function normalizeArray(items, schema, options) {
    const results = [];
    const allErrors = [];
    for (let i = 0; i < items.length; i++) {
        const item = items[i];
        const result = normalizeSingle(item, schema, options);
        if (!result.success) {
            if (options.strict) {
                return {
                    success: false,
                    error: `Item at index ${i}: ${result.error}`,
                    missingFields: result.missingFields,
                };
            }
            allErrors.push(`[${i}] ${result.error}`);
            continue;
        }
        results.push(result.data);
    }
    return Object.assign({ success: true, data: results }, (allErrors.length > 0 ? { _warnings: allErrors } : {}));
}
// ─── Public API ───────────────────────────────────────────────────────────────
/**
 * Normalize an API response (object or array) against a schema.
 *
 * @param data    Raw API response
 * @param schema  Field mapping schema
 * @param options Normalizer options
 */
function normalize(data, schema, options = {}) {
    // Null / undefined guard
    if (data === null || data === undefined) {
        if (options.strict) {
            return {
                success: false,
                error: "Input data is null or undefined",
            };
        }
        return { success: true, data: {} };
    }
    if (Array.isArray(data)) {
        return normalizeArray(data, schema, options);
    }
    if (!(0, helpers_1.isPlainObject)(data)) {
        return {
            success: false,
            error: `Expected a plain object or array, received: ${typeof data}`,
        };
    }
    return normalizeSingle(data, schema, options);
}
/**
 * Convenience: infer a schema from a sample API response.
 * Useful for quick exploration — refine the output before using in production.
 */
var mapper_2 = require("./mapper");
Object.defineProperty(exports, "inferSchema", { enumerable: true, get: function () { return mapper_2.inferSchema; } });
