// @munesoft/logx/dist/index.cjs
'use strict';

// @munesoft/logx v1.0.0 — https://github.com/munesoft/logx
// MIT License

// ── color ────────────────────────────────────────────────────────────────────
// Zero-dependency ANSI color support
const ESC = '\x1b[';
const RESET = `${ESC}0m`;

const isTTY = () =>
  typeof process !== 'undefined' &&
  process.stdout &&
  process.stdout.isTTY &&
  process.env.NO_COLOR === undefined;

const codes = {
  reset:   [0,   0],
  bold:    [1,  22],
  dim:     [2,  22],
  // Foreground
  black:   [30, 39],
  red:     [31, 39],
  green:   [32, 39],
  yellow:  [33, 39],
  blue:    [34, 39],
  magenta: [35, 39],
  cyan:    [36, 39],
  white:   [37, 39],
  gray:    [90, 39],
  // Bright
  brightRed:    [91, 39],
  brightGreen:  [92, 39],
  brightYellow: [93, 39],
  brightBlue:   [94, 39],
  brightCyan:   [96, 39],
};

function colorize(str, ...colorNames) {
  if (!isTTY()) return str;
  const open  = colorNames.map(n => codes[n] ? `${ESC}${codes[n][0]}m` : '').join('');
  const close = colorNames.map(n => codes[n] ? `${ESC}${codes[n][1]}m` : '').join('');
  return `${open}${str}${close}`;
}

// Browser %c styles
const browserStyles = {
  info:    'color: #4ade80; font-weight: bold',
  warn:    'color: #facc15; font-weight: bold',
  error:   'color: #f87171; font-weight: bold',
  debug:   'color: #818cf8; font-weight: bold',
  default: 'color: #94a3b8; font-weight: bold',
};

export { colorize, browserStyles, isTTY, RESET };

// ── levels ───────────────────────────────────────────────────────────────────
export const LEVELS = {
  silent: 0,
  error:  1,
  warn:   2,
  info:   3,
  debug:  4,
};

export const LEVEL_LABELS = {
  error: 'error',
  warn:  'warn',
  info:  'info',
  debug: 'debug',
};

// Level display config: [symbol, colorName(s)]
export const LEVEL_STYLE = {
  info:  { symbol: '✔', colors: ['green'],        label: 'INFO'  },
  warn:  { symbol: '⚠', colors: ['yellow'],       label: 'WARN'  },
  error: { symbol: '✖', colors: ['brightRed', 'bold'], label: 'ERROR' },
  debug: { symbol: '◆', colors: ['gray'],         label: 'DEBUG' },
};

export function isLevelEnabled(currentLevel, targetLevel) {
  return LEVELS[targetLevel] <= LEVELS[currentLevel];
}

// ── utils ────────────────────────────────────────────────────────────────────
// ── Environment detection ───────────────────────────────────────────────────
export const isBrowser = typeof window !== 'undefined' && typeof window.document !== 'undefined';
export const isNode    = typeof process !== 'undefined' && process.versions != null && process.versions.node != null;

export function isDev() {
  if (isNode) {
    const env = process.env.NODE_ENV || 'development';
    return env !== 'production';
  }
  // Browser: check common dev indicators
  if (isBrowser) {
    return window.location && (
      window.location.hostname === 'localhost' ||
      window.location.hostname === '127.0.0.1' ||
      window.location.hostname.endsWith('.local')
    );
  }
  return true;
}

// ── Safe JSON serialization ──────────────────────────────────────────────────
export function safeStringify(obj, indent) {
  const seen = new WeakSet();
  return JSON.stringify(obj, (_key, value) => {
    if (typeof value === 'object' && value !== null) {
      if (seen.has(value)) return '[Circular]';
      seen.add(value);
    }
    if (typeof value === 'bigint') return value.toString() + 'n';
    if (value instanceof Error) return serializeError(value);
    return value;
  }, indent);
}

export function serializeError(err) {
  return {
    name:    err.name,
    message: err.message,
    stack:   err.stack,
    ...(err.code  ? { code:  err.code  } : {}),
    ...(err.cause ? { cause: err.cause } : {}),
  };
}

// ── Inline sprintf-style formatting ─────────────────────────────────────────
export function sprintf(msg, ...args) {
  if (!args.length || typeof msg !== 'string') return msg;
  let i = 0;
  return msg.replace(/%([sdioO%])/g, (match, type) => {
    if (type === '%') return '%';
    const arg = args[i++];
    if (arg === undefined) return match;
    switch (type) {
      case 's': return String(arg);
      case 'd':
      case 'i': return parseInt(arg, 10);
      case 'o':
      case 'O': return safeStringify(arg);
      default:  return arg;
    }
  });
}

// ── Flatten extra args into a data object ───────────────────────────────────
export function extractData(args) {
  let data   = {};
  let errors = [];

  for (const arg of args) {
    if (arg instanceof Error) {
      errors.push(arg);
    } else if (arg !== null && typeof arg === 'object' && !Array.isArray(arg)) {
      Object.assign(data, arg);
    }
  }

  return { data, errors };
}

// ── Timestamp ────────────────────────────────────────────────────────────────
export function timestamp() {
  return Math.floor(Date.now() / 1000);
}

export function isoTimestamp() {
  return new Date().toISOString();
}

// ── Format key=value pairs for pretty output ────────────────────────────────
export function formatKV(data) {
  return Object.entries(data)
    .map(([k, v]) => {
      if (typeof v === 'object' && v !== null) {
        v = safeStringify(v);
      }
      return `${k}=${v}`;
    })
    .join(' ');
}

// ── format ───────────────────────────────────────────────────────────────────
// ── Pretty formatter (Node terminal) ────────────────────────────────────────
export function formatPretty(entry) {
  const { level, message, data, errors, context, ts } = entry;
  const style  = LEVEL_STYLE[level] || LEVEL_STYLE.info;
  const color  = (...args) => colorize(...args);

  const useColor = isTTY();

  // Symbol + message
  const symbol = useColor
    ? colorize(style.symbol, ...style.colors)
    : style.symbol;

  const msg = useColor
    ? colorize(message, 'white')
    : message;

  // Merge context + data for kv display
  const allData = { ...context, ...data };
  const kvStr   = Object.keys(allData).length
    ? '  ' + (useColor ? colorize(formatKV(allData), 'gray') : formatKV(allData))
    : '';

  // Timestamp (dim)
  const timeStr = useColor
    ? colorize(isoTimestamp(), 'dim', 'gray')
    : isoTimestamp();

  let line = `${symbol} ${msg}${kvStr}`;

  // Error stack traces
  const stackLines = errors
    .map(e => {
      const header = useColor
        ? colorize(`  ${e.name}: ${e.message}`, 'brightRed')
        : `  ${e.name}: ${e.message}`;
      const stack = (e.stack || '')
        .split('\n')
        .slice(1)
        .map(l => (useColor ? colorize(l, 'gray') : l))
        .join('\n');
      return `${header}\n${stack}`;
    })
    .join('\n');

  return stackLines ? `${line}\n${stackLines}` : line;
}

// ── JSON formatter ───────────────────────────────────────────────────────────
export function formatJSON(entry) {
  const { level, message, data, errors, context, ts } = entry;

  const out = {
    level,
    message,
    timestamp: ts,
    ...context,
    ...data,
  };

  if (errors.length === 1) {
    out.error = serializeError(errors[0]);
  } else if (errors.length > 1) {
    out.errors = errors.map(serializeError);
  }

  return safeStringify(out);
}

// ── Browser formatter ────────────────────────────────────────────────────────
export function formatBrowser(entry) {
  const { level, message, data, errors, context } = entry;
  const style = LEVEL_STYLE[level] || LEVEL_STYLE.info;

  const styles = {
    info:  'color: #4ade80; font-weight: bold',
    warn:  'color: #facc15; font-weight: bold',
    error: 'color: #f87171; font-weight: bold',
    debug: 'color: #818cf8; font-weight: bold',
  };

  const allData = { ...context, ...data };
  const hasMeta = Object.keys(allData).length > 0 || errors.length > 0;

  // Returns [formatStr, ...substitutions, ...extras]
  const parts = [
    `%c${style.symbol} %c${message}`,
    styles[level] || 'color: #94a3b8',
    'color: inherit; font-weight: normal',
  ];

  if (hasMeta) {
    parts.push(allData);
  }

  if (errors.length) {
    errors.forEach(e => parts.push(e));
  }

  return parts;
}

// ── core ─────────────────────────────────────────────────────────────────────
// ── Default config ───────────────────────────────────────────────────────────
const defaultConfig = {
  level:     isDev() ? 'debug' : 'info',
  pretty:    true,
  json:      false,
  timestamp: true,
  silent:    false,
};

// ── Global state ─────────────────────────────────────────────────────────────
let _config     = { ...defaultConfig };
let _filters    = [];
let _transports = [];
let _asyncCtx   = {};   // simple global context (Node AsyncLocalStorage not needed for basic usage)

// ── Write to output ──────────────────────────────────────────────────────────
function write(entry) {
  if (_config.silent) return;
  if (!isLevelEnabled(_config.level, entry.level)) return;

  // Run filters
  for (const filter of _filters) {
    if (!filter(entry.level, entry)) return;
  }

  // Run custom transports (they can suppress default output by returning false)
  let suppress = false;
  for (const transport of _transports) {
    const result = transport(entry);
    if (result === false) suppress = true;
  }
  if (suppress) return;

  if (isBrowser) {
    const parts = formatBrowser(entry);
    const method = entry.level === 'error' ? 'error'
                 : entry.level === 'warn'  ? 'warn'
                 : entry.level === 'debug' ? 'debug'
                 : 'log';
    console[method](...parts);
  } else {
    const out = _config.json ? formatJSON(entry) : formatPretty(entry);
    const method = entry.level === 'error' ? 'error'
                 : entry.level === 'warn'  ? 'warn'
                 : 'log';
    console[method](out);
  }
}

// ── Build a log entry ────────────────────────────────────────────────────────
function buildEntry(level, msg, args, context = {}) {
  // Handle sprintf-style formatting
  const message = sprintf(msg, ...args.filter(a => typeof a !== 'object' || a instanceof Error ? true : false).filter(a => !(a instanceof Error)));

  const { data, errors } = extractData(args);

  return {
    level,
    message,
    data,
    errors,
    context: { ..._asyncCtx, ...context },
    ts: _config.timestamp ? timestamp() : undefined,
  };
}

// ── Core log function ────────────────────────────────────────────────────────
function log(msg, ...args) {
  if (typeof msg === 'object' && msg !== null && !Array.isArray(msg)) {
    // logx({ silent: true }) style config shorthand
    if (msg.silent === true) { _config.silent = true; return; }
  }
  const entry = buildEntry('info', String(msg), args);
  write(entry);
}

// ── Named level methods ──────────────────────────────────────────────────────
log.info  = (msg, ...args) => write(buildEntry('info',  String(msg), args));
log.warn  = (msg, ...args) => write(buildEntry('warn',  String(msg), args));
log.error = (msg, ...args) => write(buildEntry('error', String(msg), args));
log.debug = (msg, ...args) => write(buildEntry('debug', String(msg), args));

// ── JSON mode ────────────────────────────────────────────────────────────────
log.json = (msg, ...args) => {
  const prevJson = _config.json;
  _config.json = true;
  write(buildEntry('info', String(msg), args));
  _config.json = prevJson;
};

// ── Config ───────────────────────────────────────────────────────────────────
log.config = (opts = {}) => {
  _config = { ..._config, ...opts };
};

// ── Child logger (inherited context) ─────────────────────────────────────────
log.child = (context = {}) => {
  function childLog(msg, ...args) {
    write(buildEntry('info', String(msg), args, context));
  }

  for (const level of ['info', 'warn', 'error', 'debug']) {
    childLog[level] = (msg, ...args) =>
      write(buildEntry(level, String(msg), args, context));
  }

  childLog.json = (msg, ...args) => {
    const prevJson = _config.json;
    _config.json = true;
    write(buildEntry('info', String(msg), args, context));
    _config.json = prevJson;
  };

  childLog.child = (extraCtx = {}) => log.child({ ...context, ...extraCtx });

  return childLog;
};

// ── Async context ─────────────────────────────────────────────────────────────
log.context = (ctx = {}) => {
  _asyncCtx = { ..._asyncCtx, ...ctx };
};

// ── Filtering ────────────────────────────────────────────────────────────────
log.filter = (fn) => {
  _filters.push(fn);
};

// ── Transports ───────────────────────────────────────────────────────────────
log.use = (transport) => {
  _transports.push(transport);
};

// ── Grouping ─────────────────────────────────────────────────────────────────
log.group = (label, fn) => {
  if (typeof console.group === 'function') {
    console.group(label);
    try { fn(); } finally { console.groupEnd(); }
  } else {
    console.log(`── ${label} ──`);
    fn();
  }
};

// ── Reset (useful for tests) ─────────────────────────────────────────────────
log._reset = () => {
  _config     = { ...defaultConfig };
  _filters    = [];
  _transports = [];
  _asyncCtx   = {};
};

export { log };
export default log;


// CJS exports
Object.defineProperty(exports, '__esModule', { value: true });
exports.default = log;
exports.log     = log;
module.exports  = log;
module.exports.default = log;
module.exports.log     = log;
