// @munesoft/memoryx/dist/index.js
// src/adapters/memory.ts
var MemoryAdapter = class {
  constructor() {
    this.store = /* @__PURE__ */ new Map();
  }
  get(id) {
    return this.store.get(id);
  }
  set(entry) {
    this.store.set(entry.id, entry);
  }
  delete(id) {
    return this.store.delete(id);
  }
  clear() {
    this.store.clear();
  }
  all() {
    return Array.from(this.store.values());
  }
  size() {
    return this.store.size;
  }
};

// src/adapters/file.ts
var FileAdapter = class {
  constructor(path = "./.memoryx.json") {
    this.store = /* @__PURE__ */ new Map();
    this.writeTimer = null;
    this.fs = null;
    this.loaded = false;
    this.path = path;
  }
  async ensureLoaded() {
    if (this.loaded) return;
    this.loaded = true;
    try {
      this.fs = await import("fs");
    } catch {
      return;
    }
    try {
      const raw = this.fs.readFileSync(this.path, "utf8");
      const parsed = JSON.parse(raw);
      for (const entry of parsed) this.store.set(entry.id, entry);
    } catch {
    }
  }
  scheduleFlush() {
    if (!this.fs) return;
    if (this.writeTimer) clearTimeout(this.writeTimer);
    this.writeTimer = setTimeout(() => {
      this.flushSync();
    }, 50);
  }
  flushSync() {
    if (!this.fs) return;
    try {
      const data = JSON.stringify(Array.from(this.store.values()));
      this.fs.writeFileSync(this.path, data, "utf8");
    } catch {
    }
  }
  async get(id) {
    await this.ensureLoaded();
    return this.store.get(id);
  }
  async set(entry) {
    await this.ensureLoaded();
    this.store.set(entry.id, entry);
    this.scheduleFlush();
  }
  async delete(id) {
    await this.ensureLoaded();
    const ok = this.store.delete(id);
    if (ok) this.scheduleFlush();
    return ok;
  }
  async clear() {
    await this.ensureLoaded();
    this.store.clear();
    this.scheduleFlush();
  }
  async all() {
    await this.ensureLoaded();
    return Array.from(this.store.values());
  }
  async size() {
    await this.ensureLoaded();
    return this.store.size;
  }
};

// src/adapters/redis.ts
var RedisAdapter = class {
  constructor({ client, prefix = "memoryx" }) {
    this.client = client;
    this.prefix = prefix;
  }
  key(id) {
    return `${this.prefix}:${id}`;
  }
  async get(id) {
    const raw = await this.client.get(this.key(id));
    if (!raw) return void 0;
    try {
      return JSON.parse(raw);
    } catch {
      return void 0;
    }
  }
  async set(entry) {
    await this.client.set(this.key(entry.id), JSON.stringify(entry));
  }
  async delete(id) {
    const removed = await this.client.del(this.key(id));
    return removed > 0;
  }
  async clear() {
    const keys = await this.client.keys(`${this.prefix}:*`);
    if (keys.length === 0) return;
    await this.client.del(keys);
  }
  async all() {
    const keys = await this.client.keys(`${this.prefix}:*`);
    const out = [];
    for (const k of keys) {
      const raw = await this.client.get(k);
      if (!raw) continue;
      try {
        out.push(JSON.parse(raw));
      } catch {
      }
    }
    return out;
  }
  async size() {
    const keys = await this.client.keys(`${this.prefix}:*`);
    return keys.length;
  }
};

// src/adapters/factory.ts
function createAdapter(option) {
  if (!option || option === "memory") {
    return new MemoryAdapter();
  }
  if (option === "file") {
    return new FileAdapter();
  }
  if (option === "redis") {
    throw new Error(
      "memoryx: store: 'redis' requires a client. Use store: { type: 'redis', client }"
    );
  }
  if (typeof option === "object" && option !== null) {
    if (typeof option.get === "function" && typeof option.set === "function" && typeof option.all === "function") {
      return option;
    }
    const cfg = option;
    if (cfg.type === "memory") return new MemoryAdapter();
    if (cfg.type === "file") return new FileAdapter(cfg.path);
    if (cfg.type === "redis") {
      if (!cfg.client) {
        throw new Error("memoryx: redis store requires a 'client' field");
      }
      return new RedisAdapter({
        client: cfg.client,
        prefix: cfg.prefix
      });
    }
    throw new Error(`memoryx: unknown adapter type '${cfg.type}'`);
  }
  throw new Error(`memoryx: invalid store option`);
}

// src/utils/uid.ts
var ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
function getRandomBytes(n) {
  const buf = new Uint8Array(n);
  const g = typeof globalThis !== "undefined" ? globalThis : {};
  if (g.crypto && typeof g.crypto.getRandomValues === "function") {
    g.crypto.getRandomValues(buf);
    return buf;
  }
  for (let i = 0; i < n; i++) buf[i] = Math.random() * 256 | 0;
  return buf;
}
function uid() {
  const t = Date.now().toString(36);
  const bytes = getRandomBytes(10);
  let r = "";
  for (let i = 0; i < bytes.length; i++) {
    r += ALPHABET[bytes[i] % 62];
  }
  return `${t}-${r}`;
}

// src/utils/time.ts
var now = () => Date.now();

// src/utils/text.ts
var STOP_WORDS = /* @__PURE__ */ new Set([
  "a",
  "an",
  "and",
  "are",
  "as",
  "at",
  "be",
  "but",
  "by",
  "do",
  "for",
  "from",
  "has",
  "have",
  "i",
  "if",
  "in",
  "is",
  "it",
  "its",
  "me",
  "my",
  "no",
  "not",
  "of",
  "on",
  "or",
  "so",
  "that",
  "the",
  "this",
  "to",
  "was",
  "were",
  "will",
  "with",
  "you",
  "your"
]);
var TOKEN_RE = /[a-z0-9]+/gi;
function toText(data) {
  if (data == null) return "";
  if (typeof data === "string") return data;
  if (typeof data === "number" || typeof data === "boolean" || typeof data === "bigint") {
    return String(data);
  }
  try {
    return JSON.stringify(data);
  } catch {
    return String(data);
  }
}
function tokenize(text) {
  if (!text) return [];
  const lower = text.toLowerCase();
  const matches = lower.match(TOKEN_RE);
  if (!matches) return [];
  const out = [];
  const seen = /* @__PURE__ */ new Set();
  for (const t of matches) {
    if (t.length < 2 && !/^\d$/.test(t)) continue;
    if (STOP_WORDS.has(t)) continue;
    if (seen.has(t)) continue;
    seen.add(t);
    out.push(t);
  }
  return out;
}
function tokenizeQuery(query) {
  if (!query) return [];
  const lower = query.toLowerCase();
  const matches = lower.match(TOKEN_RE);
  if (!matches) return [];
  return matches.filter((t) => !STOP_WORDS.has(t));
}
function approxTokens(text) {
  if (!text) return 0;
  return Math.ceil(text.length / 4);
}
function truncateToBudget(text, tokenBudget) {
  const charBudget = tokenBudget * 4;
  if (text.length <= charBudget) return text;
  return text.slice(0, charBudget - 1) + "\u2026";
}

// src/recall/fuzzy.ts
function jaccard(a, b) {
  if (a.length === 0 && b.length === 0) return 0;
  if (a.length === 0 || b.length === 0) return 0;
  const setA = new Set(a);
  const setB = new Set(b);
  let intersection = 0;
  for (const t of setA) if (setB.has(t)) intersection++;
  const union = setA.size + setB.size - intersection;
  return union === 0 ? 0 : intersection / union;
}
function levenshtein(a, b) {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;
  let prev = new Array(b.length + 1);
  let curr = new Array(b.length + 1);
  for (let j = 0; j <= b.length; j++) prev[j] = j;
  for (let i = 1; i <= a.length; i++) {
    curr[0] = i;
    for (let j = 1; j <= b.length; j++) {
      const cost = a.charCodeAt(i - 1) === b.charCodeAt(j - 1) ? 0 : 1;
      curr[j] = Math.min(
        curr[j - 1] + 1,
        prev[j] + 1,
        prev[j - 1] + cost
      );
    }
    [prev, curr] = [curr, prev];
  }
  return prev[b.length];
}
function tokenSimilarity(a, b) {
  if (a === b) return 1;
  const longer = a.length >= b.length ? a : b;
  if (longer.length === 0) return 1;
  const dist = levenshtein(a, b);
  return 1 - dist / longer.length;
}

// src/recall/score.ts
var FUZZY_THRESHOLD = 0.7;
function scoreRelevance(query, entry) {
  if (!query) return 0;
  const q = query.toLowerCase().trim();
  if (!q) return 0;
  let substring = 0;
  if (entry.text.includes(q)) {
    substring = Math.min(0.6, 0.3 + q.length * 0.02);
  }
  const queryTokens = tokenizeQuery(q);
  if (queryTokens.length === 0) {
    return substring;
  }
  const jac = jaccard(queryTokens, entry.tokens);
  let fuzzyHits = 0;
  for (const qt of queryTokens) {
    let best = 0;
    for (const et of entry.tokens) {
      const sim = tokenSimilarity(qt, et);
      if (sim > best) best = sim;
      if (best === 1) break;
    }
    if (best >= FUZZY_THRESHOLD) fuzzyHits += best;
  }
  const fuzzy = queryTokens.length > 0 ? fuzzyHits / queryTokens.length : 0;
  const blended = Math.max(
    substring,
    jac * 0.7 + fuzzy * 0.3
  );
  return Math.min(1, blended);
}

// src/ranking/rank.ts
var HOUR = 60 * 60 * 1e3;
var DAY = 24 * HOUR;
function recencyScore(entry, t = now()) {
  const age = Math.max(0, t - entry.accessedAt);
  const halfLife = 7 * DAY;
  return Math.pow(0.5, age / halfLife);
}
function importanceScore(entry) {
  const i = entry.importance;
  if (typeof i !== "number" || Number.isNaN(i)) return 0.5;
  if (i < 0) return 0;
  if (i > 1) return 1;
  return i;
}
function combinedScore(relevance, recency, importance) {
  return relevance * 0.65 + recency * 0.2 + importance * 0.15;
}

// src/core/summarize.ts
var extractiveSummarizer = (entries) => {
  if (entries.length === 0) return "";
  if (entries.length === 1) return entryToLine(entries[0]);
  const scored = entries.map((e) => ({
    entry: e,
    score: e.importance * Math.log(1 + e.hits + 1)
  })).sort((a, b) => b.score - a.score);
  const keepCount = Math.min(10, Math.max(3, Math.ceil(entries.length / 2)));
  const kept = scored.slice(0, keepCount);
  const lines = [];
  const seen = /* @__PURE__ */ new Set();
  for (const { entry } of kept) {
    const line = entryToLine(entry);
    const key = line.toLowerCase().slice(0, 40);
    if (seen.has(key)) continue;
    seen.add(key);
    lines.push(line);
  }
  return lines.join(" \u2022 ");
};
function entryToLine(entry) {
  if (typeof entry.data === "string") return entry.data;
  return entry.text;
}

// src/core/memory.ts
var DEFAULT_OPTS = {
  ai: false,
  shortTerm: true,
  longTerm: true,
  session: true,
  namespace: "default",
  shortTermLimit: 100,
  longTermLimit: 1e3,
  contextBudget: 2e3,
  dedupe: true,
  dedupeThreshold: 0.85,
  compress: true
};
var Memory = class {
  constructor(options = {}) {
    this.opts = { ...DEFAULT_OPTS, ...options };
    this.adapter = createAdapter(options.store);
    this.hooks = options.hooks ?? {};
    this.summarizer = options.summarizer ?? extractiveSummarizer;
  }
  // ─── remember ──────────────────────────────────────────────────────────────
  /**
   * Store a value in memory.
   *
   * @param data    Anything JSON-serializable.
   * @param options Per-entry options — ttl, tags, importance, namespace, layer.
   * @returns       The id of the stored entry.
   */
  async remember(data, options = {}) {
    const t = now();
    const text = toText(data);
    const tokens = tokenize(text);
    const layer = options.layer ?? "short";
    const namespace = options.namespace ?? this.opts.namespace;
    const entry = {
      id: options.id ?? uid(),
      data,
      text,
      tokens,
      createdAt: t,
      accessedAt: t,
      hits: 0,
      importance: clamp01(options.importance ?? 0.5),
      layer,
      namespace,
      tags: options.tags ?? [],
      expiresAt: options.ttl ? t + options.ttl : null
    };
    if (this.opts.dedupe) {
      const merged = await this.tryMerge(entry);
      if (merged) {
        await this.callHook(() => this.hooks.onRemember?.(merged));
        return merged.id;
      }
    }
    await this.adapter.set(entry);
    await this.enforceLimits();
    await this.callHook(() => this.hooks.onRemember?.(entry));
    return entry.id;
  }
  /**
   * Stream-style append — store partial output. Each call adds an incremental
   * chunk; the chunks share a stream id so they can later be reassembled.
   *
   * Useful for capturing streaming AI responses without losing partial content
   * if a stream is interrupted.
   */
  async stream(chunk, options = {}) {
    const streamId = options.streamId ?? "_stream";
    return this.remember(chunk, {
      ...options,
      tags: [...options.tags ?? [], `stream:${streamId}`]
    });
  }
  // ─── recall ────────────────────────────────────────────────────────────────
  /**
   * Search memory for entries matching `query`. Combines keyword, fuzzy,
   * and substring matching, then ranks by relevance + recency + importance.
   */
  async recall(query, options = {}) {
    const limit = options.limit ?? 10;
    const minScore = options.minScore ?? 0;
    const ns = options.namespace;
    const tagFilter = options.tags;
    const layerFilter = options.layer;
    await this.evictExpired();
    const all = await this.adapter.all();
    const t = now();
    const results = [];
    for (const entry of all) {
      if (entry.expiresAt && entry.expiresAt <= t) continue;
      if (ns && entry.namespace !== ns) continue;
      if (layerFilter && entry.layer !== layerFilter) continue;
      if (tagFilter && tagFilter.length > 0) {
        const overlap = tagFilter.some((tag) => entry.tags.includes(tag));
        if (!overlap) continue;
      }
      const relevance = scoreRelevance(query, entry);
      const recency = recencyScore(entry, t);
      const importance = importanceScore(entry);
      const score = combinedScore(relevance, recency, importance);
      if (score < minScore) continue;
      if (query && relevance === 0) continue;
      results.push({
        entry,
        data: entry.data,
        score,
        scores: { relevance, recency, importance }
      });
    }
    if (options.recentFirst) {
      results.sort((a, b) => b.entry.accessedAt - a.entry.accessedAt);
    } else {
      results.sort((a, b) => b.score - a.score);
    }
    const sliced = results.slice(0, limit);
    for (const r of sliced) {
      r.entry.hits++;
      r.entry.accessedAt = t;
      await this.adapter.set(r.entry);
    }
    await this.callHook(() => this.hooks.onRecall?.(query, sliced));
    return sliced;
  }
  // ─── forget / clear ────────────────────────────────────────────────────────
  /**
   * Remove entries that match `query`. If `query` is empty, this is a no-op
   * (use `clear()` to wipe everything).
   */
  async forget(query) {
    if (!query) return 0;
    const matches = await this.recall(query, { limit: 1e3, minScore: 0.2 });
    const ids = [];
    for (const m of matches) {
      const ok = await this.adapter.delete(m.entry.id);
      if (ok) ids.push(m.entry.id);
    }
    if (ids.length > 0) {
      await this.callHook(() => this.hooks.onForget?.(ids));
    }
    return ids.length;
  }
  /**
   * Wipe all memory. Optionally restrict to a single namespace.
   */
  async clear(namespace) {
    if (!namespace) {
      await this.adapter.clear();
      await this.callHook(() => this.hooks.onClear?.());
      return;
    }
    const all = await this.adapter.all();
    const ids = [];
    for (const e of all) {
      if (e.namespace === namespace) {
        await this.adapter.delete(e.id);
        ids.push(e.id);
      }
    }
    if (ids.length > 0) {
      await this.callHook(() => this.hooks.onForget?.(ids));
    }
  }
  // ─── compress ──────────────────────────────────────────────────────────────
  /**
   * Manually compress old entries into synthetic digests. Useful for keeping
   * token budgets low in long-running agents without waiting for the
   * automatic eviction trigger.
   *
   * ```ts
   * await memory.compress({ olderThan: '7d' });
   * ```
   *
   * Returns counts of source entries folded, digests created, and approximate
   * tokens saved.
   */
  async compress(options = {}) {
    const minBatch = options.minBatch ?? 3;
    const cutoff = options.olderThan != null ? now() - parseDuration(options.olderThan) : Number.POSITIVE_INFINITY;
    const all = await this.adapter.all();
    const groups = /* @__PURE__ */ new Map();
    for (const e of all) {
      if (e.compressed) continue;
      if (e.layer !== "long") continue;
      if (e.accessedAt > cutoff) continue;
      if (options.namespace && e.namespace !== options.namespace) continue;
      const key = e.namespace;
      const bucket = groups.get(key) ?? [];
      bucket.push(e);
      groups.set(key, bucket);
    }
    let compressed = 0;
    let digests = 0;
    let tokensSaved = 0;
    for (const [, batch] of groups) {
      if (batch.length < minBatch) continue;
      const before = batch.reduce((sum, e) => sum + approxTokens(e.text), 0);
      const digest = await this.compressBatch(batch);
      if (!digest) continue;
      for (const e of batch) await this.adapter.delete(e.id);
      const after = approxTokens(digest.text);
      compressed += batch.length;
      digests += 1;
      tokensSaved += Math.max(0, before - after);
    }
    const result = { compressed, digests, tokensSaved };
    if (compressed > 0) {
      await this.callHook(() => this.hooks.onCompress?.(result));
    }
    return result;
  }
  /**
   * Fold a batch of entries into a single synthetic digest entry.
   * Returns the digest entry (also persisted to the adapter), or null if
   * the batch was empty.
   */
  async compressBatch(batch) {
    if (batch.length === 0) return null;
    const summary = await this.summarizer(batch);
    if (!summary) return null;
    const t = now();
    const namespace = batch[0].namespace;
    const importance = batch.reduce((m, e) => Math.max(m, e.importance), 0);
    const tagSet = /* @__PURE__ */ new Set();
    for (const e of batch) for (const tag of e.tags) tagSet.add(tag);
    tagSet.add("compressed");
    const text = summary;
    const digest = {
      id: uid(),
      data: text,
      text,
      tokens: tokenize(text),
      createdAt: t,
      accessedAt: t,
      hits: 0,
      importance,
      layer: "long",
      namespace,
      tags: Array.from(tagSet),
      expiresAt: null,
      compressed: true,
      compressedFrom: batch.length
    };
    await this.adapter.set(digest);
    return digest;
  }
  // ─── context ───────────────────────────────────────────────────────────────
  /**
   * Build a prompt-ready, size-aware context string for the given scope.
   *
   * Prioritizes:
   *   - high-importance entries
   *   - recent entries (always includes the N most recent)
   *   - relevance to `scope` if it's a meaningful query string
   *
   * Returns plain text safe to splice into an LLM prompt.
   */
  async context(scope, options = {}) {
    const budget = options.budget ?? this.opts.contextBudget;
    const ns = options.namespace ?? this.opts.namespace;
    const recentCount = options.recentCount ?? 3;
    await this.evictExpired();
    const all = await this.adapter.all();
    if (all.length === 0) return "";
    const t = now();
    const filtered = all.filter((e) => {
      if (e.expiresAt && e.expiresAt <= t) return false;
      if (ns && ns !== "default" && e.namespace !== ns) return false;
      return true;
    });
    if (filtered.length === 0) return "";
    const scored = filtered.map((entry) => {
      const relevance = scope ? scoreRelevance(scope, entry) : 0;
      const recency = recencyScore(entry, t);
      const importance = importanceScore(entry);
      const score = scope ? relevance * 0.5 + importance * 0.3 + recency * 0.2 : importance * 0.6 + recency * 0.4;
      return { entry, score };
    });
    scored.sort((a, b) => b.score - a.score);
    const recent = [...filtered].sort((a, b) => b.accessedAt - a.accessedAt).slice(0, recentCount);
    const recentIds = new Set(recent.map((e) => e.id));
    const ordered = [];
    const seen = /* @__PURE__ */ new Set();
    for (const e of recent) {
      ordered.push(e);
      seen.add(e.id);
    }
    for (const { entry } of scored) {
      if (seen.has(entry.id)) continue;
      ordered.push(entry);
      seen.add(entry.id);
    }
    const lines = [];
    let used = 0;
    for (const entry of ordered) {
      const line = formatLine(entry, recentIds.has(entry.id));
      const cost = approxTokens(line) + 1;
      if (used + cost > budget) {
        if (this.opts.ai && budget - used > 20) {
          lines.push(truncateToBudget(line, budget - used));
        }
        break;
      }
      lines.push(line);
      used += cost;
    }
    return lines.join("\n");
  }
  // ─── inspection ────────────────────────────────────────────────────────────
  /** Return all stored entries (useful for debugging / persistence). */
  async all() {
    await this.evictExpired();
    return this.adapter.all();
  }
  /** Number of currently stored entries. */
  async size() {
    await this.evictExpired();
    return this.adapter.size();
  }
  // ─── internals ─────────────────────────────────────────────────────────────
  /** Evict TTL-expired entries. Cheap — runs before reads. */
  async evictExpired() {
    const t = now();
    const all = await this.adapter.all();
    for (const e of all) {
      if (e.expiresAt && e.expiresAt <= t) {
        await this.adapter.delete(e.id);
      }
    }
  }
  /**
   * If a near-duplicate exists, merge into it (bumps importance and
   * accessedAt, refreshes data). Returns the merged entry, or null if
   * nothing similar was found.
   */
  async tryMerge(entry) {
    const all = await this.adapter.all();
    let best = null;
    for (const e of all) {
      if (e.namespace !== entry.namespace) continue;
      if (e.text === entry.text) {
        best = { entry: e, sim: 1 };
        break;
      }
      const sim = jaccard(e.tokens, entry.tokens);
      if (sim >= this.opts.dedupeThreshold && (!best || sim > best.sim)) {
        best = { entry: e, sim };
      }
    }
    if (!best) return null;
    const merged = {
      ...best.entry,
      data: entry.data,
      text: entry.text,
      tokens: entry.tokens,
      tags: dedupeArray([...best.entry.tags, ...entry.tags]),
      importance: Math.max(best.entry.importance, entry.importance),
      accessedAt: now(),
      hits: best.entry.hits + 1,
      // Earliest TTL wins — prevents merging from extending lifetime.
      expiresAt: minTTL(best.entry.expiresAt, entry.expiresAt),
      merged: true
    };
    await this.adapter.set(merged);
    return merged;
  }
  /**
   * Enforce short-term and long-term layer limits.
   * Short-term overflow → promote oldest to long-term.
   * Long-term overflow  → evict lowest combined (importance + recency) score.
   */
  async enforceLimits() {
    const all = await this.adapter.all();
    const shortEntries = all.filter((e) => e.layer === "short");
    const longEntries = all.filter((e) => e.layer === "long");
    if (this.opts.shortTerm && shortEntries.length > this.opts.shortTermLimit) {
      const overflow = shortEntries.length - this.opts.shortTermLimit;
      const sorted = [...shortEntries].sort((a, b) => {
        const sa = a.importance + a.accessedAt / 1e13;
        const sb = b.importance + b.accessedAt / 1e13;
        return sa - sb;
      });
      for (let i = 0; i < overflow; i++) {
        const e = sorted[i];
        if (this.opts.longTerm) {
          await this.adapter.set({ ...e, layer: "long" });
        } else {
          await this.adapter.delete(e.id);
        }
      }
    }
    if (this.opts.longTerm && longEntries.length > this.opts.longTermLimit) {
      const overflow = longEntries.length - this.opts.longTermLimit;
      const t = now();
      const sorted = [...longEntries].sort((a, b) => {
        const sa = importanceScore(a) * 0.6 + recencyScore(a, t) * 0.4;
        const sb = importanceScore(b) * 0.6 + recencyScore(b, t) * 0.4;
        return sa - sb;
      });
      if (this.opts.compress) {
        const COMPRESS_BATCH = Math.max(overflow + 2, 4);
        const candidates = sorted.slice(0, COMPRESS_BATCH).filter((e) => !e.compressed);
        if (candidates.length >= 2) {
          await this.compressBatch(candidates);
          for (const v of candidates) await this.adapter.delete(v.id);
          const remaining = await this.adapter.all();
          const stillLong = remaining.filter((e) => e.layer === "long");
          if (stillLong.length > this.opts.longTermLimit) {
            const stillOver = stillLong.length - this.opts.longTermLimit;
            const reSorted = [...stillLong].sort((a, b) => {
              const sa = importanceScore(a) * 0.6 + recencyScore(a, t) * 0.4;
              const sb = importanceScore(b) * 0.6 + recencyScore(b, t) * 0.4;
              return sa - sb;
            });
            for (let i = 0; i < stillOver; i++) {
              await this.adapter.delete(reSorted[i].id);
            }
          }
          return;
        }
      }
      for (let i = 0; i < overflow; i++) {
        await this.adapter.delete(sorted[i].id);
      }
    }
  }
  async callHook(fn) {
    try {
      await fn();
    } catch {
    }
  }
};
function clamp01(n) {
  if (Number.isNaN(n)) return 0.5;
  if (n < 0) return 0;
  if (n > 1) return 1;
  return n;
}
function dedupeArray(arr) {
  return Array.from(new Set(arr));
}
function minTTL(a, b) {
  if (a === null) return b;
  if (b === null) return a;
  return Math.min(a, b);
}
function formatLine(entry, isRecent) {
  const marker = isRecent ? "\u2022" : "-";
  const text = typeof entry.data === "string" ? entry.data : entry.text;
  return `${marker} ${text}`;
}
function parseDuration(input) {
  if (typeof input === "number") return input;
  const m = /^(\d+(?:\.\d+)?)\s*(ms|s|m|h|d|w)$/i.exec(input.trim());
  if (!m) {
    throw new Error(`memoryx: invalid duration '${input}' \u2014 use e.g. '7d' or 60000`);
  }
  const n = parseFloat(m[1]);
  const unit = m[2].toLowerCase();
  const mult = {
    ms: 1,
    s: 1e3,
    m: 6e4,
    h: 36e5,
    d: 864e5,
    w: 6048e5
  };
  return n * mult[unit];
}

// src/index.ts
function memoryx(options = {}) {
  return new Memory(options);
}
var index_default = memoryx;
export {
  FileAdapter,
  Memory,
  MemoryAdapter,
  RedisAdapter,
  index_default as default,
  extractiveSummarizer,
  memoryx
};
