// @munesoft/objx/dist/cjs/index.js
"use strict";
/**
 * @munesoft/objx
 * A modern, fast, safe lodash.get / lodash.set / deep-merge alternative.
 *
 * @example
 * import * as objx from "@munesoft/objx";
 * objx.get(obj, "user.profile.name", "anonymous");
 * objx.set(state, "user.name", "Alice", { immutable: true });
 * objx.merge(defaults, overrides, { arrays: "concat" });
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearPathCache = exports.parsePath = exports.defaults = exports.del = exports.has = exports.merge = exports.set = exports.compile = exports.get = void 0;
var index_js_1 = require("./get/index.js");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return index_js_1.get; } });
Object.defineProperty(exports, "compile", { enumerable: true, get: function () { return index_js_1.compile; } });
var index_js_2 = require("./set/index.js");
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return index_js_2.set; } });
var index_js_3 = require("./merge/index.js");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return index_js_3.merge; } });
var has_del_js_1 = require("./utils/has-del.js");
Object.defineProperty(exports, "has", { enumerable: true, get: function () { return has_del_js_1.has; } });
Object.defineProperty(exports, "del", { enumerable: true, get: function () { return has_del_js_1.del; } });
Object.defineProperty(exports, "defaults", { enumerable: true, get: function () { return has_del_js_1.defaults; } });
var index_js_4 = require("./path/index.js");
Object.defineProperty(exports, "parsePath", { enumerable: true, get: function () { return index_js_4.parsePath; } });
Object.defineProperty(exports, "clearPathCache", { enumerable: true, get: function () { return index_js_4.clearPathCache; } });
